src/shogun/modelselection/NewGridSearch.h

```diff
@@ -0,0 +1,354 @@
+		 */
+		GridParameters* attach(const std::string& param, GridParameters* node);
+
+		/**
+		 * Attach a new SGVector at a given location in the tree
+		 * defined by the parameter name
+		 * @param name
+		 * @param node
+		 */
+		template <typename T>
+		GridParameters* attach(const std::string& param, SGVector<T> values)
+		{
+			if (!set_param_helper(param, make_any(values)))
+				SG_SERROR("Could not attach %s", param.c_str());
+			return this;
+		}
+
+	protected:
+		/**
+		 * Resets the internal state of the node.
+		 */
+		void reset();
+
+		/**
+		 * Creates a new node based on a CSGObject
+		 * @param name
+		 * @param obj
+		 */
+		void create_node(const std::string& name, CSGObject* obj) override;
+
+		/**
+		 * Gets the next ParameterNode representation
+		 * of this instance
+		 * @return
+		 */
+		ParameterNode* get_next() final;
+
+		/**
+		 * Gets the current ParameterNode representation
+		 * of this instance
+		 * @return
+		 */
+		ParameterNode* get_current() final;
+
+		bool set_param_helper(const std::string& param, const Any& value) final;
+
+		/**
+		 * Checks if all nodes below this one have been completed.
+		 *
+		 * @return
+		 */
+		bool is_complete();
+
+		/**
+		 * Checks if all nested nodes in m_nodes have been completed.
+		 * @return
+		 */
+		bool is_inner_complete(const std::string&);
+
+		/**
+		 * Set node complete status. This is needed if a parent
+		 * needs to reset the complete status of a child node.
+		 *
+		 * @param flag
+		 */
+		void set_node_complete(bool flag)
+		{
+			m_node_complete = flag;
+		}
+
+		/**
+		 * Child node complete status getter.
+		 *
+		 * @return
+		 */
+		bool check_child_node_done()
+		{
+			return m_node_complete;
+		}
+
+	private:
+		/** if it is the first iteration of the node */
+		bool m_first;
+		/** tracks current node being iterated over. A node can have several
+		 * internal nodes */
+		std::map<std::string, std::vector<std::shared_ptr<ParameterNode>>>::
+		    iterator m_current_node;
+		/** tracks current internal node being iterated over */
+		std::vector<std::shared_ptr<ParameterNode>>::iterator
+		    m_current_internal_node;
+		/** tracks current parameter being iterated over*/
+		std::map<std::string, Any>::iterator m_current_param;
+		/** current iterator of a given parameter */
+		std::unordered_map<std::string, Any> m_param_iter;
+		/** iterator begin of a given parameter */
+		std::unordered_map<std::string, Any> m_param_begin;
+		/** iterator end of a given parameter */
+		std::unordered_map<std::string, Any> m_param_end;
+		/** whether the node is complete, i.e. all parameter combinations have
+		 * been generated */
+		bool m_node_complete;
+	};
+
+	class GridSearch : public GridParameters
+	{
+	public:
+		explicit GridSearch(CSGObject* model) : GridParameters(model)
+		{
+		}
+
+		const char* get_name() const noexcept override
+		{
+			return "GridSearch";
+		}
+
+		void train(CFeatures* data);
+
+	private:
+		SGVector<float64_t> m_scores;
```

Comment:
need to think about evaluation and how to store each model's performance

Reply:
Check out the modelselection code we have. There is a criterion, which is either maximised or minimised. And then there is an "evaluation" object that determines how the instantiated models are evaluated (xvalidation, etc).
I think this modular setup is quite important

* modelselection (how are the parameters to be tried generated? grid-search? gradient search?)
* model (train, apply)
* criterion (accuracy? AUC?)
* evaluation (how is the model evaluated? xvalidation? test set? marginal likelihood? theoretical measures? other?)
* parameters (these are the best parameters found, or some intermediate best results)